### ISSUE TITLE
Part: Regression with boolean common

### ISSUE THREAD

**Issue by wwmayer** [FreeCAD/FreeCAD#17689]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem description

Boolean intersection (common) used to be a commutative operation, i.e. the order of input shapes has no impact on the result. With the current implementation this is not longer the case. Furthermore, for more than 2 input shapes the result in general is wrong.

Steps:
* Create a part box, cylinder and sphere
* Select the box, cylinder and sphere and do a boolean common
* Select the cylinder, sphere and box and do a boolean common
* Select the sphere, box and cylinder and do a boolean common
* Compare the results: they are all different and they are all wrong

Or in a more formal way:
The result of the input shapes S1, S2, S3, ... Sn used to be S1 ∧ S2 ∧ S3 ... ∧Sn while now it is (S1 ∧ S2) ∨(S1 ∧ S3) ∨ ...∨(S1 ∧ Sn)

### Full version info

```shell
OS: Ubuntu 22.04.5 LTS (XFCE/xubuntu/xcb)
Architecture: x86_64
Version: 1.1.0dev.39128 (Git)
Build type: debug
Branch: main
Hash: 4c50f2944287ab966a063ef4091480e65ac99acd
Python 3.10.12, Qt 5.15.3, Coin 4.0.0, Vtk 7.1.1, OCC 7.5.1
Locale: German/Germany (de_DE)
Stylesheet/Theme/QtStyle: unset/Classic theme/Qt default
```

### Subproject(s) affected?

Part

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by luzpaz** [comment]
Any idea what code change was responsible for regression ?

**Comment by wwmayer** [comment]
The way how `TopoShape::makeElementBoolean` handles argument and tool shapes doesn't work with Common. The issue is caused by https://github.com/FreeCAD/FreeCAD/pull/14182

**Comment by AIRCAP** [comment]
This also fails when a boolean common is done with a compound.

**Comment by AIRCAP** [comment]
I suggest introducing a BooleanSetOperation as a new base class

**Comment by wwmayer** [comment]
> At this point it would be better to get rid of the MultiFusion as a separate implementation

That's not possible without breaking many existing project files.

A way to handle this could be to override the Build() method of `FCBRepAlgoAPI_Common`.

**Comment by kadet1090** [comment]
So, the problem here is that we are right now aligned to the way that OCC understands that operation - BrepAlgoApi_Common computes common shape between `arguments` and `tools`, so basically the same as if we would first make a compound of elements and then call compound betwen first shape and the rest. 

Problem here is that we have 1.0 that works with the same behavior and so if we now change it to be like it used to be we risk that we break files created in 1.0. So we would need to support two paths and somehow differentiate between them. That would be a nightmare to maintain

**Comment by kadet1090** [comment]
So I see it this way:

This issue cannot be blocker - we don't have any example of file that is actually broken by that change, and it seems that we are more closely aligned to OCC right now. Depending on use both behaviors probably make sense and so a group of people will argue that if they are computing common of shape A with B,C,D they want to have A and (B or C or D) - another group of people will want to have common of all of these shapes so A and B and C and D.

My proposal is to update the API to expose enum flag that would allow us to choose wanted behavior. We can then expose this as property of boolean operation for users to choose. For files created before 1.0 we can then set that property to be aligned with older behavior and for newer files we can do current one (to preserve compatibility with 1.0).

**Comment by DeepSOIC** [comment]
> Problem here is that we have 1.0 that works with the same behavior and so if we now change it to be like it used to be we risk that we break files created in 1.0.

yes, but a) that's already happened on 0.20->1.0 transition; b) it may be aligned with how occ works, but the way it works now makes no sense. And i highly doubt anyone is relying on this behavior, simply because it's pointless. Also, i hope we don't call every bug a feature now.

If we need an independent multicommon implementation, we can simply clone Part XOR and modify it very slightly to produce a proper multicommon. Or even closer, see Macro Overlap.

**Comment by kadet1090** [comment]
@DeepSOIC 

> a) that's already happened on 0.20->1.0

Can you elaborate on that? Yes, there is a break of compatibility between 0.21 and 1.0 which is not great to say the least - which should be fixed. That does not however change a fact, that changing the behavior now would introduce break of compatibility between 1.0 -> 1.1 which makes it significantly harder than saying to simply restore old behavior.

> b) it may be aligned with how occ works, but the way it works now makes no sense. And i highly doubt anyone is relying on this behavior, simply because it's pointless. Also, i hope we don't call every bug a feature now.

What is pointless what is not depends on usecase and a model, if something works in a certain way even if we think that it is incorrect way the users will depend on it. We had at least few issues where someone did depend on behavior that was clearly bugous (using part design features outside of body) and we removed it. Essentially every bug if it stays for long enough is a feature because someone uses it. 

Now we have two conflicting interpretations of how that feature should work and still no proof of broken existing files or many screaming users - which suggests that impact of that particular bug is not that big. It still has `Priority: High` label and `Type: Regression` so it's not like anyone wants to say that we don't want to fix it - simply given that it is how it works in current stable version being 1.0 there is no reason to block 1.1 from being released because it's not fixed.

As mentioned in second part of the comment - my proposal is basically to add parameter that would allow us to use one approach or the other depending on circumstances. The old one should be the default for Part (as indeed - there there is no separation between these "arguments" and "tools") and the parameter should be probably hidden. That way files created with 0.21 and lower will still work fine, files from version where the current behavior was present will also work. 

For Part Design which uses the same method to compute the result I am however not sure if the old behavior makes more sense - Part Design has clear separation between arguments (being the body) and tools (being well, selected tools) is clear and so the result of operation "Compute common of my body and X, Y, Z" seems to be more like `Body and (X or Y or Z)` like - so in such case we should simply expose UI element that allows user to pick the implementation they want.

**Comment by DeepSOIC** [comment]
> we don't have any example of file that is actually broken by that change,

I'm pretty sure i have some... after all, i have bumped into this issue when making something with a purpose. But i couldn't easily find one, so i made my own:

[7-sided-allen-ball-head_uses-multicommon.zip](https://github.com/user-attachments/files/24624674/7-sided-allen-ball-head_uses-multicommon.zip)

<img width="1050" height="786" alt="Image" src="https://github.com/user-attachments/assets/07075f13-c3d5-423c-99aa-3e6cfb70edc2" />

**Comment by DeepSOIC** [comment]
> What is pointless what is not depends on usecase and a model

i agree, but the current behavior does not align with the name "Common" in my mind. It's such a bizarre operation i'm having a hard time coming up with a name for it. "Slurp" maybe.

**Comment by DeepSOIC** [comment]
> As mentioned in second part of the comment - my proposal is basically to add parameter that would allow us to use one approach or the other depending on circumstances.

i'm for sure ok with that. Sorry for ranting.

**Comment by kadet1090** [comment]
> Sorry for ranting.

At least we know that there is someone who really cares about this issue 😅 

If anyone would like to tackle that problem before I have time to do it here is an example of PR that introduces such compatibility measure: https://github.com/FreeCAD/FreeCAD/pull/26141 and here is another one https://github.com/FreeCAD/FreeCAD/pull/26781
